File: code/class-02-classes-and-objects/car.py
"""
car.py

Class 02 – Classes and Objects

Demonstrates:
  - Defining a class with private-by-convention attributes
  - Multiple constructors via default parameters
  - The use of `self`
  - __str__ / __repr__ overrides
"""


class Car:
    """Represents a car with brand, model, year, and price."""

    def __init__(
        self,
        brand: str  = "Unknown",
        model: str  = "Unknown",
        year: int   = 0,
        price: float = 0.0,
    ) -> None:
        self.brand = brand
        self.model = model
        self.year  = year
        self.price = price

    # Behaviour
    def display_info(self) -> None:
        print(self)

    def __str__(self) -> str:
        return f"{self.year} {self.brand} {self.model} – ${self.price:,.2f}"

    def __repr__(self) -> str:
        return (
            f"Car(brand={self.brand!r}, model={self.model!r}, "
            f"year={self.year}, price={self.price})"
        )


# ── Demo ─────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    car1 = Car("Toyota", "Corolla", 2024, 22_000.00)
    car2 = Car("Honda",  "Civic",   2023, 20_500.00)
    car3 = Car()  # default values

    car1.display_info()
    car2.display_info()
    car3.display_info()

    # Identity vs Equality
    car_a = Car("Toyota", "Corolla", 2024, 22_000.00)
    print(f"\ncar1 is car_a         : {car1 is car_a}")          # False
    print(f"car1 == car_a (default): {car1 == car_a}")           # False (no __eq__ override)

File: code/class-02-classes-and-objects/Car.java
/**
 * Car.java
 *
 * Class 02 – Classes and Objects
 *
 * Demonstrates:
 *   - Defining a class with private fields
 *   - Multiple constructors (overloading)
 *   - The `this` keyword
 *   - A toString() override
 */
public class Car {

    // Fields
    private String brand;
    private String model;
    private int    year;
    private double price;

    // Default constructor
    public Car() {
        this("Unknown", "Unknown", 0, 0.0);
    }

    // Parameterised constructor
    public Car(String brand, String model, int year, double price) {
        this.brand = brand;
        this.model = model;
        this.year  = year;
        this.price = price;
    }

    // Getters
    public String getBrand() { return brand; }
    public String getModel() { return model; }
    public int    getYear()  { return year;  }
    public double getPrice() { return price; }

    // Behaviour
    public void displayInfo() {
        System.out.println(this); // calls toString()
    }

    @Override
    public String toString() {
        return year + " " + brand + " " + model + " – $" + String.format("%.2f", price);
    }

    // ----- Demo -----
    public static void main(String[] args) {
        Car car1 = new Car("Toyota", "Corolla", 2024, 22_000.00);
        Car car2 = new Car("Honda",  "Civic",   2023, 20_500.00);
        Car car3 = new Car(); // default values

        car1.displayInfo();
        car2.displayInfo();
        car3.displayInfo();

        // Identity vs Equality
        Car carA = new Car("Toyota", "Corolla", 2024, 22_000.00);
        System.out.println("car1 == carA  : " + (car1 == carA));           // false (different refs)
        System.out.println("car1.equals(carA): " + car1.equals(carA));     // false (no equals override)
    }
}
